src/ra_utility

```diff
@@ -0,0 +1,241 @@
+
+
+_NODISCARD _CONSTANT_FN operator""_ba(_In_ unsigned long long n) noexcept
+{
+	return static_cast<ByteAddress>(n);
+} // end operator""_ba
+
+
+_NODISCARD _CONSTANT_FN operator""_achid(_In_ unsigned long long n) noexcept {
+	return static_cast<AchievementID>(n);
+} // end operator""_achid
+
+
+_NODISCARD _CONSTANT_FN operator""_lbid(_In_ unsigned long long n) noexcept {
+	return static_cast<LeaderboardID>(n);
+} // end operator""_lbid
+
+
+_NODISCARD _CONSTANT_FN operator""_gameid(_In_ unsigned long long n) noexcept {
+	return static_cast<GameID>(n);
+} // end operator""_gameid
+
+} // inline namespace int_literals
+
+template<typename SignedType, class = std::enable_if_t<std::is_signed_v<SignedType>>> _NODISCARD _CONSTANT_FN
+to_unsigned(_In_ SignedType st) noexcept { return static_cast<std::make_unsigned_t<SignedType>>(st); }
+
+template<typename UnsignedType, class = std::enable_if_t<std::is_unsigned_v<UnsignedType>>> _NODISCARD _CONSTANT_FN
+to_signed(_In_ UnsignedType st) noexcept { return static_cast<std::make_signed_t<UnsignedType>>(st); }
+
+#if _WIN32
+using float_type = float;
+#elif _WIN64
+using float_type = double;
+#else
+using float_type = long double; // should we even care bout this?
+#endif // _WIN32
+
+
+template<typename Arithmetic, class = std::enable_if_t<std::is_arithmetic_v<Arithmetic>>> _NODISCARD tstring 
+to_tstring(_In_ Arithmetic a) noexcept
+{
+#if _MBCS
+	return std::to_string(a);
+#elif _UNICODE
+	return std::to_wstring(a);
+#else
+#error Unknown character set detected, only MultiByte and Unicode character sets are supported!
+#endif // UNICODE
+} // end function to_tstring
+
+template<typename Integral, class = std::enable_if_t<std::is_integral_v<Integral>>> _NODISCARD _CONSTANT_FN
+to_floating(_In_ Integral i) noexcept { return static_cast<float_type>(i); }
+
+template<typename FloatingPoint, class = std::enable_if_t<std::is_floating_point_v<FloatingPoint>>>
+_NODISCARD _CONSTANT_FN ftoi(_In_ FloatingPoint fp) noexcept
+{ 
+	if(std::signbit(fp))
+		return std::lround(fp); 
+	return to_unsigned(std::lround(fp));
+}
+
+template<typename Arithmetic, class = std::enable_if_t<std::is_arithmetic_v<Arithmetic>>> _NODISCARD _CONSTANT_FN
+sqr(_In_ Arithmetic a) noexcept { return std::pow(a, 2); }
+
+template<typename Enum, typename = std::enable_if_t<std::is_enum_v<Enum>>> _NODISCARD _CONSTANT_VAR
+etoi(_In_ Enum e) noexcept { return static_cast<std::underlying_type_t<Enum>>(e); }
+
+// function alias template for etoi (EnumToIntegral)
+template<typename Enum> _NODISCARD _CONSTANT_VAR to_integral = etoi<Enum>;
+
+namespace detail {
+
+/// <summary>
+///   Casts <typeparamref name="InputString" /> into an
+///   <typeparamref name="OutputString" />. This won't work with multi-byte
+///   strings, only byte strings. It can be used to convert <c>std::string</c> to
+///   and from <c>std::wstring</c> but that's only if they are byte strings.
```

Comment:
any data sent to/from the server is assumed to be UTF-8. as such, any internal strings should also be assumed to be UTF-8. `Widen` and `Narrow` _must_ function with UTF-8 strings.

Reply:
This will be gone if that's the case.